### PR TITLE
[FLINK-21413][state] Clean TtlMapState and TtlListState after all elements are expired

### PR DESCRIPTION
**Pull request by Jiayi-Liao**
## What is the purpose of the change

Solve the memory leak in #TtlMapState and #TtlListState with #FsStateBackend and incremental cleanup feature. 

## Brief change log

Changes in #TtlMapState and #TtlListState  

* In #getUnexpiredOrNull method, return null if all the elements are expired

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by Jiayi-Liao** [comment]
@carp84 It took me some time to dig how to test this change, but unfortunately there may not be an easy way to achieve this. 

Take the `TtlListState` as an example, in `TtlStateTestBase`, Flink uses `TtlStateTestContextBase.isOriginalEmptyValue()` to test whether the state is cleared, which is `Objects.equals(emptyValue, getOriginal());` in TtlListState's testing. To test my change, I need to verify the result of `AbstractHeapAppendingState.getInternal()`, but unit testing uses `MockInternalKvState.getInternal()`, which never returns null in `TtlListState`.

**Comment by carp84** [comment]
Sorry for the late response @Jiayi-Liao that I'm a little bit overloaded these days and may need some more time to get back, will ping back once ready.

**Comment by carp84** [comment]
@Jiayi-Liao please check e840893 in my [private branch](https://github.com/carp84/flink/tree/flink-21413) about adding the unit test (I've confirmed the added case would fail w/o the fix here). I'd also suggest to refine the fix in the way of 256c160. Please let me know your thoughts.

**Comment by Jiayi-Liao** [comment]
> @Jiayi-Liao please check [e840893](https://github.com/apache/flink/commit/e84089332fef495e70324c67f9d9a7297c927065) in my [private branch](https://github.com/carp84/flink/tree/flink-21413) about adding the unit test (I've confirmed the added case would fail w/o the fix here). I'd also suggest to refine the fix in the way of [256c160](https://github.com/apache/flink/commit/256c16072e50a569e824cf79115f80234d833c31). Please let me know your thoughts.

Your implementation looks good to me, I tried to verify my logic in #TtlStateTestContextBase#isOriginalEmptyValue before, which makes it hard to implement.

Thanks for the help. I've refined my commit and cherry-pick your commit of unit testing here. Could you take another look? @carp84

**Review by carp84 — APPROVED** [review]
Thanks for the follow up @Jiayi-Liao , the latest changes LGTM.

However, since I'm a co-author now, it would be better to involve some other committers to review / double check.

@Myasuka @klion26 @rkhachatryan @StephanEwen could you help to also take a look here? Thanks.

**Comment by StephanEwen** [comment]
To make a review here, I would need a bit more context about this problem. @Jiayi-Liao if you can give me a bit of context for you assumptions on this fix, I can do a review here.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/ttl/TtlStateTestBase.java
Comment (Myasuka) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/ttl/TtlStateTestBase.java:492]: I think the kernel logic is to verify that current `StateTable` at `currentNamespace` is null and current logic seems a bit strange as its name `is original null` hided what we want.

I prefer to change this line to
~~~java
assertTrue("Original state table should be cleared", isCurrentStateTableNull());
~~~

And we could also adding method below:
~~~java
    @SuppressWarnings({"rawtypes", "unchecked"})
    private boolean isCurrentStateTableNull() {
        InternalKvState<?, String, ?> original = ((AbstractTtlDecorator<InternalKvState<?, String, ?>>) ctx.ttlState).original;
        Preconditions.checkState(original instanceof AbstractHeapState);
        return ((AbstractHeapState) original).getStateTable().get(ctx.currentNamespace) == null;
    }
~~~

By doing so, we could remove those strange `isValueOfCurrentKeyNull` and `isOriginalNull` methods.

**Review by Myasuka — COMMENTED** [review]
@Jiayi-Liao Thanks for your contribution, please take a look at my comments.

**Comment by Jiayi-Liao** [comment]
> To make a review here, I would need a bit more context about this problem. @Jiayi-Liao if you can give me a bit of context for you assumptions on this fix, I can do a review here.

There is a memory leak issue in the incremental cleanup of #TtlMapState and #TtlListState in FsStateBackend. The #TtlIncrementalCleanup expires the state from #StateTable when #getUnexpiredOrNull(ttlValue) returns null, but #TtlMapState and #TtlListState never return null in the method, which means the #TtlMapState and #TtlListState will never be removed from #StateTable. We met the issue when we're trying to process a stream of new devices and the range of new devices is unbounded.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/ttl/TtlStateTestBase.java
Comment (carp84) [reply flink-runtime/src/test/java/org/apache/flink/runtime/state/ttl/TtlStateTestBase.java:492]: I agree it is somewhat over-designed here but the original thought is to try not to fail the test if `assumeTrue(incrementalCleanupSupported())` returns `true` for RocksDB backend in the future.

The suggested way works but the name `isCurrentStateTableNull` is misleading... It's "the state of the mapping for the composite of active key and given namespace" (referring to `StateTable#get(N namespace)`) that should be null, rather than the current state table. `isOriginalCleared` may be a better name.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlListState.java
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlListState.java:108]: Can this ever happen here? Are we wrapping the `UserFacingListState` which maps `null` to an empty list, or are empty lists never possible?

**Review by StephanEwen — APPROVED** [review]
Change looks good to me.

@carp84 and @Myasuka have good comments for the tests, please take those into account.

As a side note, the TTL code looks like it could use some performance improvements. For example there seems to be a lot of extra things done that could be avoided:
  - More `System.currentTimeMillis` calls than necessary
  - Defensive copies within cleanup iteration (should be avoidable with the right abstraction)
  - Extra list object creation
 Let's file a follow-up issue for these improvements.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/ttl/TtlStateTestBase.java
Comment (Jiayi-Liao) [reply flink-runtime/src/test/java/org/apache/flink/runtime/state/ttl/TtlStateTestBase.java:492]: Here we're testing the existence of the state of specific key and namespace, right? so why not rename the method to `isOriginalExisting()`, which can also keep the same style with another method `isOriginalEmptyValue()`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlListState.java
Comment (Jiayi-Liao) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlListState.java:108]: This could happen before this change(after the elements are all expired). So... if a job restores from a previous Flink version without this change, it may still happen.

**Comment by Jiayi-Liao** [comment]
@carp84 @Myasuka Sorry for the late reply(pretty busy these days). I've improved the codes based on your comments, please take another look, thanks!

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/ttl/TtlStateTestBase.java
Comment (Myasuka) [reply flink-runtime/src/test/java/org/apache/flink/runtime/state/ttl/TtlStateTestBase.java]: My original point is that we could avoid to introduce those `isValueOfCurrentKeyNull` related methods by just introduce one single `isCurrentStateTableNull` method within `TtlStateTestBase.java` (no matter what name it is). 

FLINK-21413 is about to fix a performance problem to avoid many unnecessary empty hash map and state entry, instead of incorrect query result. This new added test should only targets for heap-based keyed state backend and we can safely cast to `AbstractHeapState` here.

Comment (Jiayi-Liao) [reply flink-runtime/src/test/java/org/apache/flink/runtime/state/ttl/TtlStateTestBase.java]: @Myasuka  Oh.. I see what you mean here. Very good point. :) 

I've improved the codes based on your comments.

**Review by Myasuka — APPROVED** [review]
LGTM, and we can merge it once CI turned green.

**Comment by Jiayi-Liao** [comment]
The error of CI should be unrelated. 

```
Mar 29 07:53:23 Error: Could not find or load main class org.apache.flink.kubernetes.entrypoint.KubernetesSessionClusterEntrypoint

Mar 29 07:53:41 [FAIL] 'Run kubernetes session test (default input)' failed after 1 minutes and 59 seconds! Test exited with exit code 1
```

**Comment by carp84** [comment]
We need to get a green build to commit the change, let me trigger another pre-commit run, will merge it once the build passes.

**Comment by carp84** [comment]
@flinkbot run azure

**Comment by carp84** [comment]
@flinkbot run azure

**Comment by carp84** [comment]
I guess something is wrong with the machine running the azure build, and re-trigger the test against last build will always run on the same machine, so I've merged the commits, rebased the latest master branch, and force-pushed to trigger a new one. Let's see how it goes this time.
> 2021-03-30T04:37:46.3354934Z Mar 30 04:37:45 sed: couldn't open temporary file /opt/flink/conf/sedMH11PH: Read-only file system
> 2021-03-30T04:37:46.3355705Z Mar 30 04:37:45 sed: couldn't open temporary file /opt/flink/conf/sed3FBMoH: Read-only file system
> 2021-03-30T04:37:46.3356478Z Mar 30 04:37:45 /docker-entrypoint.sh: line 73: /opt/flink/conf/flink-conf.yaml: Read-only file system
> 2021-03-30T04:37:46.3357550Z Mar 30 04:37:45 sed: couldn't open temporary file /opt/flink/conf/sedmsw7OJ: Read-only file system

**Pull request closed without merging**
